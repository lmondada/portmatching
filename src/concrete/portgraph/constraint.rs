//! Constraints for port graph matching.
//!
//! There are two main logic bits
//!  - how to express several constraints as mutually exclusive
//!    constraints; this can be found in the mutex submodule.
//!  - how to decompose a constraint into a list of constraints. Note that this
//!    has to be consistent with how the indexing scheme binds key-values.

mod mutex;

use std::{cmp, collections::BTreeSet, iter};

use itertools::Itertools;
use petgraph::visit::EdgeCount;
use portgraph::{NodeIndex, PortGraph, PortView};
use thiserror::Error;

use crate::{
    utils::{portgraph::line_partition, sort_with_indices},
    Constraint, HashMap,
};

use super::{indexing::PGIndexKey, predicate::PGPredicate};
use mutex::*;

/// A constraint on a port graph.
pub type PGConstraint<W = ()> = Constraint<PGIndexKey, PGPredicate<W>>;

impl PartialOrd for PGConstraint {
    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for PGConstraint {
    fn cmp(&self, other: &Self) -> cmp::Ordering {
        self.max_key().cmp(&other.max_key())
    }
}

impl PGConstraint {
    /// The key to use for comparisons.
    ///
    /// Constraints are sorted by their largest key binding.
    fn max_key(&self) -> (PGIndexKey, &PGPredicate) {
        let max_key = self
            .required_bindings()
            .iter()
            .max()
            .copied()
            .unwrap_or(PGIndexKey::PathRoot { index: 0 });
        (max_key, self.predicate())
    }
}

#[cfg(test)]
mod tests {
    use portgraph::PortOffset;
    use rstest::{fixture, rstest};

    use super::*;

    fn vname(root_id: usize, root_port: Option<PortOffset>, node_index: usize) -> PGIndexKey {
        if let Some(root_port) = root_port {
            PGIndexKey::AlongPath {
                path_root: root_id,
                path_start_port: root_port,
                path_length: node_index,
            }
        } else {
            assert_eq!(root_port, None);
            PGIndexKey::PathRoot { index: root_id }
        }
    }

    fn filter(
        left_port: PortOffset,
        (left_root_id, left_root_port, left_node_index): (usize, Option<PortOffset>, usize),
        right_port: PortOffset,
        (right_root_id, right_root_port, right_node_index): (usize, Option<PortOffset>, usize),
    ) -> PGConstraint {
        PGConstraint::try_new(
            PGPredicate::IsConnected {
                left_port,
                right_port,
            }
            .into(),
            vec![
                vname(left_root_id, left_root_port, left_node_index),
                vname(right_root_id, right_root_port, right_node_index),
            ],
        )
        .unwrap()
    }

    #[fixture]
    fn constraints() -> Vec<PGConstraint> {
        vec![
            filter(
                PortOffset::new_outgoing(0),
                (1, None, 0),
                PortOffset::new_incoming(0),
                (0, Some(PortOffset::new_incoming(0)), 1),
            ),
            filter(
                PortOffset::new_incoming(0),
                (2, None, 0),
                PortOffset::new_outgoing(0),
                (0, None, 0),
            ),
            filter(
                PortOffset::new_incoming(0),
                (0, Some(PortOffset::new_outgoing(0)), 2),
                PortOffset::new_outgoing(0),
                (0, Some(PortOffset::new_incoming(0)), 1),
            ),
            filter(
                PortOffset::new_incoming(0),
                (0, Some(PortOffset::new_outgoing(0)), 2),
                PortOffset::new_outgoing(1),
                (0, Some(PortOffset::new_incoming(0)), 1),
            ),
        ]
    }

    #[rstest]
    fn test_to_mut_ex_assign_0(constraints: Vec<PGConstraint>) {
        let tree = PGPredicate::to_constraints_tree(constraints.clone());
        assert_eq!(
            tree.children(tree.root())
                .map(|(_, c)| c.clone())
                .collect_vec(),
            [
                filter(
                    PortOffset::new_incoming(0),
                    (0, Some(PortOffset::new_outgoing(0)), 2),
                    PortOffset::new_outgoing(0),
                    (0, Some(PortOffset::new_incoming(0)), 1)
                ),
                filter(
                    PortOffset::new_incoming(0),
                    (0, Some(PortOffset::new_outgoing(0)), 2),
                    PortOffset::new_outgoing(1),
                    (0, Some(PortOffset::new_incoming(0)), 1),
                ),
            ]
        );
    }

    #[rstest]
    fn test_to_mut_ex_assign_1(constraints: Vec<PGConstraint>) {
        let tree = PGPredicate::to_constraints_tree(constraints[..2].to_vec());
        assert_eq!(
            tree.children(tree.root())
                .map(|(_, c)| c.clone())
                .collect_vec(),
            [filter(
                PortOffset::new_outgoing(0),
                (1, None, 0),
                PortOffset::new_incoming(0),
                (0, Some(PortOffset::new_incoming(0)), 1),
            ),]
        );
    }

    #[test]
    fn test_constraint_ordering() {
        let a = filter(
            PortOffset::new_incoming(0),
            (0, Some(PortOffset::new_outgoing(0)), 1),
            PortOffset::new_outgoing(0),
            (0, None, 0),
        );
        let b = filter(
            PortOffset::new_incoming(0),
            (0, Some(PortOffset::new_outgoing(0)), 1),
            PortOffset::new_outgoing(0),
            (1, None, 0),
        );
        // For same literal, an AssignPredicate should be smaller
        assert!(a < b);
    }
}
