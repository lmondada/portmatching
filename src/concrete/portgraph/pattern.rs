//! Patterns for matching port graphs.

use std::iter;

use super::{
    indexing::PGIndexKey, predicate::PGTag, ConstraintSelector, PGConstraint, PGPredicate,
};
use crate::{
    constraint::{ConstraintPattern, PartialConstraintPattern},
    utils::{is_connected, portgraph::line_partition},
    HashMap, Pattern,
};

use itertools::Itertools;
use petgraph::visit::EdgeCount;
use portgraph::{LinkView, NodeIndex, PortGraph, PortView};
use thiserror::Error;

/// A concrete port graph pattern.
///
/// This has no variables, and thus can be represented by a port graph directly.
#[derive(Clone, PartialEq, Eq, Debug)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct PGPattern<G> {
    pub(super) graph: G,
    pub(super) root: Option<NodeIndex>,
}

impl Pattern for PGPattern<PortGraph> {
    type PartialPattern = PartialConstraintPattern<PGIndexKey, PGPredicate>;
    type Error = PGPatternError;

    // Type aliases
    type Key = PGIndexKey;
    type Predicate = PGPredicate;
    type Tag = PGTag;
    type Evaluator = ConstraintSelector;

    fn required_bindings(&self) -> Vec<Self::Key> {
        // TODO: this is not very efficient...
        self.clone()
            .try_into_constraint_pattern()
            .unwrap()
            .constraints()
            .iter()
            .flat_map(|c| c.required_bindings())
            .copied()
            .unique()
            .collect()
    }

    fn try_into_partial_pattern(self) -> Result<Self::PartialPattern, PGPatternError> {
        Ok(self.try_into_constraint_pattern()?.into())
    }
}

/// Converting a pattern to constraints requires setting a root first.
#[derive(Debug)]
pub struct NoRootFound;

impl<G> PGPattern<G> {
    /// Whether a root has been set.
    pub fn is_root_set(&self) -> bool {
        self.root.is_some()
    }

    /// Get the graph defining the pattern.
    pub fn graph(&self) -> &G {
        &self.graph
    }

    /// Set the pattern root.
    pub fn set_root(&mut self, root: NodeIndex) {
        self.root = Some(root);
    }

    /// Construct a pattern from a port graph, without setting root.
    pub fn from_host(graph: G) -> Self {
        Self { graph, root: None }
    }

    /// Construct a pattern from a port graph, use the default root choice.
    pub fn from_host_pick_root(graph: G) -> Self
    where
        G: LinkView,
    {
        let mut p = Self { graph, root: None };
        p.pick_root().unwrap();
        p
    }

    /// Construct a pattern given a port graph and node as root.
    pub fn from_host_with_root(graph: G, root: NodeIndex) -> Self {
        Self {
            graph,
            root: Some(root),
        }
    }
}

impl PGPattern<PortGraph> {
    /// Convert the pattern to a [`ConstraintPattern`].
    pub fn try_into_constraint_pattern(
        self,
    ) -> Result<ConstraintPattern<PGIndexKey, PGPredicate>, PGPatternError> {
        Ok(ConstraintPattern::from_constraints(constraint_vec(
            &self.graph,
            self.root.ok_or(PGPatternError::NoRoot)?,
        )?))
    }
}

impl<G: LinkView> PGPattern<G> {
    /// Whether the pattern is connected.
    ///
    /// A pattern must always be connected for pattern matching.
    pub fn is_connected(&self) -> bool {
        is_connected(&self.graph)
    }

    /// Attempt to pick a valid root.
    pub fn pick_root(&mut self) -> Result<NodeIndex, NoRootFound> {
        let root = self.graph.nodes_iter().next().ok_or(NoRootFound)?;
        self.root = Some(root);
        Ok(root)
    }
}

/// Error type for pattern generation.
#[derive(Debug, Clone, Copy, Error)]
pub enum PGPatternError {
    /// No root node was provided.
    #[error("No root node was provided")]
    NoRoot,
}

fn constraint_vec(graph: &PortGraph, root: NodeIndex) -> Result<Vec<PGConstraint>, PGPatternError> {
    if graph.edge_count() == 0 {
        return Ok(vec![PGConstraint::try_new(
            PGPredicate::HasNodeWeight(()),
            vec![PGIndexKey::root(0)],
        )
        .unwrap()]);
    }
    let mut constraints = Vec::new();
    let mut node_to_key = HashMap::from_iter([(root, PGIndexKey::root(0))]);
    let mut node_to_root_ind = HashMap::from_iter([(root, 0)]);
    for line in line_partition(graph, root) {
        let root = graph.port_node(line[0].0).unwrap();
        let n_roots = node_to_root_ind.len();
        let root_index = *node_to_root_ind.entry(root).or_insert(n_roots);
        let root_offset = graph.port_offset(line[0].0).unwrap();
        for (i, (left, right)) in line.into_iter().enumerate() {
            let left_node = graph.port_node(left).unwrap();
            let left_port = graph.port_offset(left).unwrap();
            let right_node = graph.port_node(right).unwrap();
            let right_port = graph.port_offset(right).unwrap();
            let left_key = *node_to_key.get(&left_node).expect("unknown edge LHS");
            let right_key = if let Some(right_key) = node_to_key.get(&right_node) {
                *right_key
            } else {
                // Create a new key
                let key = PGIndexKey::AlongPath {
                    path_root: root_index,
                    path_start_port: root_offset,
                    path_length: i + 1,
                };
                // Ensure it does not clash with previously bound keys
                let args = iter::once(key)
                    .chain(node_to_key.values().copied())
                    .collect_vec();
                constraints.push(
                    PGConstraint::try_new(
                        PGPredicate::IsNotEqual {
                            n_other: node_to_key.len(),
                        },
                        args,
                    )
                    .unwrap(),
                );
                node_to_key.insert(right_node, key);
                key
            };
            constraints.push(
                PGConstraint::try_new(
                    PGPredicate::IsConnected {
                        left_port,
                        right_port,
                    },
                    vec![left_key, right_key],
                )
                .unwrap(),
            );
        }
    }
    if constraints.is_empty() {
        // We add one (dummy) constraint for the empty pattern, forcing
        // the matcher to bind the first character to a position in the
        // string when matched. An alternative would be to explicitly
        // disallow empty patterns.
        constraints.push(
            PGConstraint::try_new(
                PGPredicate::IsNotEqual { n_other: 0 },
                vec![PGIndexKey::root(0)],
            )
            .unwrap(),
        );
    }
    Ok(constraints)
}
