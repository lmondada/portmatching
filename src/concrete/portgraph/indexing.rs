//! The indexing scheme for port graphs.
//!
//! This binds every node in a port graph to a name, that we use as a
//! key during pattern matching.

use std::{cmp, fmt::Debug, iter};

use portgraph::{LinkView, NodeIndex, PortGraph, PortIndex, PortOffset, PortView};

use crate::{
    indexing::{IndexedData, Key, Value},
    HashMap, IndexingScheme,
};

use super::root_candidates::find_root_candidates;

/// The indexing scheme for port graphs.
///
/// The keys are fixed relative to a set of root nodes and a path starting
/// in that root node. Here, a "path" is a very specific type of graph path:
/// at every node that the path goes through, the path must come in through
/// `PortOffset::Incoming(i)` and leave through the opposite port, i.e.
/// `PortOffset::Outgoing(i)`.
///
/// This indexing scheme can thus be thought as the equivalent of the string
/// indexing scheme (where a key is obtained by the distance from the reference
/// point, the first character), except that the key must also specify which
/// one of the reference points (the roots) it refers to.
#[derive(Clone, Default, Debug)]
pub struct PGIndexingScheme;

/// The key for the indexing scheme
#[derive(Clone, Copy, PartialEq, Eq, Hash)]
pub enum PGIndexKey {
    /// A key for a root node
    PathRoot {
        /// The index of the root node
        index: usize,
    },
    /// A key for a node along a path, given respective to a root node
    AlongPath {
        /// The root node it is referencing to
        path_root: usize,
        /// The start direction of the path at the root
        path_start_port: PortOffset,
        /// The length of the path from the root to this node
        path_length: usize,
    },
}

impl IndexingScheme for PGIndexingScheme {
    type BindMap = HashMap<PGIndexKey, NodeIndex>;

    fn required_bindings(&self, key: &Key<Self>) -> Vec<Key<Self>> {
        match *key {
            PGIndexKey::PathRoot { index } => {
                if index == 0 {
                    vec![]
                } else {
                    vec![PGIndexKey::PathRoot { index: index - 1 }]
                }
            }
            PGIndexKey::AlongPath { path_root, .. } => {
                vec![PGIndexKey::PathRoot { index: path_root }]
            }
        }
    }
}

impl IndexedData for PortGraph {
    type IndexingScheme = PGIndexingScheme;

    fn list_bind_options(
        &self,
        key: &PGIndexKey,
        known_bindings: &<PGIndexingScheme as IndexingScheme>::BindMap,
    ) -> Vec<NodeIndex> {
        if let Some(val) = known_bindings.get(key) {
            return vec![val.clone()];
        }
        match *key {
            PGIndexKey::PathRoot { index: 0 } => self.nodes_iter().collect(),
            PGIndexKey::PathRoot { index } => {
                if known_bindings
                    .get(&PGIndexKey::PathRoot { index: index - 1 })
                    .is_none()
                {
                    vec![]
                } else {
                    find_root_candidates(self, known_bindings)
                }
            }
            PGIndexKey::AlongPath {
                path_root,
                path_start_port,
                path_length,
            } => {
                let path_root = PGIndexKey::PathRoot { index: path_root };
                let Some(&root_binding) = known_bindings.get(&path_root) else {
                    return vec![];
                };
                walk_path_nodes(self, root_binding, path_start_port)
                    .nth(path_length)
                    .into_iter()
                    .collect()
            }
        }
    }
}

impl<M> IndexedData for (&PortGraph, &M) {
    type IndexingScheme = PGIndexingScheme;

    fn list_bind_options(
        &self,
        key: &Key<Self::IndexingScheme>,
        known_bindings: &<Self::IndexingScheme as IndexingScheme>::BindMap,
    ) -> Vec<Value<Self::IndexingScheme>> {
        self.0.list_bind_options(key, known_bindings)
    }
}

impl PGIndexKey {
    /// The i-th root key
    pub fn root(index: usize) -> Self {
        PGIndexKey::PathRoot { index }
    }

    fn cmp_key(&self) -> (usize, Option<PortOffset>, usize) {
        match self {
            PGIndexKey::PathRoot { index } => (*index, None, 0),
            PGIndexKey::AlongPath {
                path_root,
                path_start_port,
                path_length,
            } => (*path_root, Some(*path_start_port), *path_length),
        }
    }
}
impl PartialOrd for PGIndexKey {
    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for PGIndexKey {
    fn cmp(&self, other: &Self) -> cmp::Ordering {
        self.cmp_key().cmp(&other.cmp_key())
    }
}

impl Debug for PGIndexKey {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::PathRoot { index } => write!(f, "Root({index})"),
            Self::AlongPath {
                path_root,
                path_start_port,
                path_length,
            } => {
                let port = match path_start_port {
                    PortOffset::Incoming(i) => format!("in{i}"),
                    PortOffset::Outgoing(i) => format!("out{i}"),
                };
                write!(f, "Along({path_root}, {port})@{path_length})")
            }
        }
    }
}

/// Traverse a path, returning an iterator over the nodes and ports on the path
///
/// The triples returned are the incoming port at a node, the node and the
/// outgoing port at a node. The first incoming port is always `None`.
///
/// If `start_offset` is a PortOffset::Incoming port, then the path is traversed
/// in reverse, and thus the first port in the triple will be an outgoing port
/// and the last port will be an incoming port.
pub(super) fn walk_path<'g>(
    graph: &'g PortGraph,
    start: NodeIndex,
    start_offset: PortOffset,
) -> impl Iterator<Item = (Option<PortIndex>, NodeIndex, Option<PortIndex>)> + 'g {
    let mut next_port = graph.port_index(start, start_offset);
    iter::once((None, start, next_port)).chain(iter::from_fn(move || {
        let prev_port = graph.port_link(next_port?);
        let curr = graph.port_node(prev_port?).unwrap();
        if curr == start {
            return None;
        }
        let next_port_offset = match graph.port_offset(prev_port?).unwrap() {
            PortOffset::Incoming(offset) => PortOffset::Outgoing(offset),
            PortOffset::Outgoing(offset) => PortOffset::Incoming(offset),
        };
        next_port = graph.port_index(curr, next_port_offset);
        Some((prev_port, curr, next_port))
    }))
}

pub(super) fn walk_path_nodes<'g>(
    graph: &'g PortGraph,
    start: NodeIndex,
    start_offset: PortOffset,
) -> impl Iterator<Item = NodeIndex> + 'g {
    walk_path(graph, start, start_offset).map(|(_, node, _)| node)
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use itertools::Itertools;
    use portgraph::{LinkMut, PortMut};
    use rstest::rstest;
    use std::hash::Hash;

    use crate::HashMap;

    use super::*;

    #[test]
    fn test_walk_path() {
        let mut graph = PortGraph::new();
        let nodes = (0..4).map(|_| graph.add_node(1, 1)).collect_vec();
        nodes.iter().tuple_windows().for_each(|(&src, &dst)| {
            graph.link_nodes(src, 0, dst, 0).unwrap();
        });
        let path = walk_path_nodes(&graph, nodes[0], PortOffset::Outgoing(0));
        assert_eq!(path.collect_vec(), nodes);
    }

    fn from_edges<V: Hash + Eq>(edges: &[(V, usize, V, usize)]) -> PortGraph {
        let mut graph = PortGraph::new();
        let mut v_to_nodes = HashMap::default();
        for (src, src_port, dst, dst_port) in edges {
            let src_node = *v_to_nodes
                .entry(src)
                .or_insert_with(|| graph.add_node(0, 0));
            let n_out_ports = graph.num_inputs(src_node).max(src_port + 1);
            graph.set_num_ports(src_node, graph.num_inputs(src_node), n_out_ports, |_, _| {});
            let dst_node = *v_to_nodes
                .entry(dst)
                .or_insert_with(|| graph.add_node(0, 0));
            let n_in_ports = graph.num_outputs(dst_node).max(dst_port + 1);
            graph.set_num_ports(dst_node, n_in_ports, graph.num_outputs(dst_node), |_, _| {});
            graph
                .link_nodes(src_node, *src_port, dst_node, *dst_port)
                .unwrap();
        }
        graph
    }

    /// Assign all possible bindings to `graph`
    fn all_bindings(graph: &PortGraph) -> Vec<HashMap<PGIndexKey, NodeIndex>> {
        let mut final_bindings = vec![];
        let mut curr_bindings = vec![(0, HashMap::default())];

        while let Some((root_index, bindings)) = curr_bindings.pop() {
            let root_key = PGIndexKey::PathRoot { index: root_index };
            let candidates = graph.list_bind_options(&root_key, &bindings);
            if candidates.is_empty() {
                final_bindings.push(bindings);
                continue;
            }
            for root_binding in candidates {
                let mut new_bindings = bindings.clone();
                new_bindings.insert(root_key, root_binding);
                for port in graph.all_port_offsets(root_binding) {
                    let path_len = walk_path_nodes(graph, root_binding, port).count();
                    for path_length in 1..path_len {
                        let key = PGIndexKey::AlongPath {
                            path_root: root_index,
                            path_start_port: port,
                            path_length,
                        };
                        let candidates = graph.list_bind_options(&key, &new_bindings);
                        let binding = candidates.into_iter().exactly_one().unwrap();
                        new_bindings.insert(key, binding);
                    }
                }
                curr_bindings.push((root_index + 1, new_bindings));
            }
        }
        final_bindings
    }

    #[rstest]
    #[case(1, vec![(0, 0, 1, 0), (1, 0, 2, 0), (2, 0, 3, 0)])]
    #[case(2, vec![(0, 0, 1, 0), (0, 1, 2, 0), (3, 0, 0, 0), (3, 1, 4, 0)])]
    #[case(3, vec![(0, 0, 1, 0), (2, 0, 0, 0), (1, 0, 2, 0)])]
    #[case(4, vec![(0, 0, 1, 0), (2, 0, 0, 0), (2, 1, 1, 1)])]
    fn test_assign_bindings(
        #[case] case_id: usize,
        #[case] edges: Vec<(usize, usize, usize, usize)>,
    ) {
        let graph = from_edges(&edges);
        assert_debug_snapshot!(format!("assign_bindings_{case_id}"), all_bindings(&graph));
    }
}
