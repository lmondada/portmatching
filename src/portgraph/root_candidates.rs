//! Logic to find potential root assignments given a port graph and known bindings.
//!
//! This can be done fairly arbirarily, as long as it is consistent. This whole
//! module is hilariously inefficient, we will see when this becomes the bottleneck.
//!
//! ## Main idea
//! Every time a new root should be assign, we build a spanning tree of all
//! the previous roots that have been assigned. We can then figure out where
//! the spanning tree can be extended to add new roots.

use itertools::Itertools;
use portgraph::{NodeIndex, PortGraph, PortIndex, PortOffset, PortView};

use crate::{HashMap, HashSet};

use super::indexing::{walk_path, PGIndexKey};

pub(super) fn find_root_candidates(
    graph: &PortGraph,
    bindings: &HashMap<PGIndexKey, NodeIndex>,
) -> Vec<NodeIndex> {
    let tree = RootSpanningTree::new(graph, bindings);

    let mut root_candidates = Vec::new();

    for root in tree.nodes {
        // For each root, we can use the paths leaving ports larger than any
        // we have used for previous roots (rationale: otherwise we'd have
        // chosen that root first)
        let max_offset_used = root
            .neighbours
            .iter()
            .filter(|(_, v)| matches!(v, NeighbourType::KnownRoot(_)))
            .map(|(&offset, _)| offset)
            .max();
        root_candidates.extend(root.neighbours.into_iter().filter_map(|(k, v)| {
            if Some(k) <= max_offset_used {
                return None;
            }
            if let NeighbourType::NewRoot(new_root) = v {
                return Some(new_root);
            }
            None
        }));
    }
    root_candidates
}

#[derive(Clone, Debug, Default)]
struct RootSpanningNode {
    neighbours: HashMap<PortOffset, NeighbourType>,
}

#[derive(Clone, Debug)]
enum NeighbourType {
    Parent,
    KnownRoot(usize),
    NewRoot(NodeIndex),
}

/// A spanning tree of all the assigned roots in the graph.
///
/// At each node store where following the path along each of its ports leads
/// to. It may lead to either i) the root we come from, ii) another assigned
/// root or iii) a node with free ports that we can extend the tree at.
///
/// We ignore back-edges, i.e. paths that lead to a root already in the spanning tree.
#[derive(Clone, Debug)]
struct RootSpanningTree {
    nodes: Vec<RootSpanningNode>,
}

impl RootSpanningTree {
    fn new(graph: &PortGraph, bindings: &HashMap<PGIndexKey, NodeIndex>) -> Self {
        let known_roots = (0..)
            .map(|i| bindings.get(&PGIndexKey::PathRoot { index: i }).map(|v| *v))
            .while_some()
            .collect_vec();
        let known_roots_inv: HashMap<_, _> = known_roots.iter().copied().zip(0..).collect();
        let known_nodes = bindings.iter().map(|(_, &v)| v).collect_vec();
        let nodes_with_free_ports = nodes_with_free_ports(graph, bindings);

        let mut tree_nodes = vec![RootSpanningNode::default(); known_roots.len()];
        let mut seen_roots = HashSet::default();
        for (i, &node) in known_roots.iter().enumerate() {
            seen_roots.insert(i);
            for port in graph.all_port_offsets(node) {
                let neighbour_type = traverse_path_neighbour_type(
                    graph,
                    node,
                    port,
                    i,
                    &mut seen_roots,
                    &known_nodes,
                    &known_roots_inv,
                    &nodes_with_free_ports,
                );
                if let Some(neighbour_type) = neighbour_type {
                    tree_nodes[i].neighbours.insert(port, neighbour_type);
                }
            }
        }
        Self { nodes: tree_nodes }
    }
}

/// Traverse the path starting in (`node`, `port`) and return the type of
/// root found on the path.
///
/// More precisely:
///  - if the path leads to a smaller root, return `NeighbourType::Parent`.
///  - if the path leads to a known, larger root, return `NeighbourType::KnownRoot`
///  - otherwise, return the first potential new root along the path as
///    `NeighbourType::NewRoot`.
///
/// We handle path cycle by picking the direction of traversal that starts in a
/// smaller offset as "finding a larger root", and the other direction as "finding
/// a smaller root".
fn traverse_path_neighbour_type(
    graph: &PortGraph,
    node: NodeIndex,
    port: PortOffset,
    current_root: usize,
    seen_roots: &mut HashSet<usize>,
    known_nodes: &[NodeIndex],
    known_roots_inv: &HashMap<NodeIndex, usize>,
    nodes_with_free_ports: &HashSet<NodeIndex>,
) -> Option<NeighbourType> {
    let mut path = Vec::new();
    for (incoming_p, node, _) in walk_path(graph, node, port).skip(1) {
        if !known_nodes.contains(&node) {
            break;
        }
        let incoming_offset = incoming_p.map(|p| graph.port_offset(p).unwrap());
        if let Some(&root) = known_roots_inv.get(&node) {
            if root < current_root {
                return Some(NeighbourType::Parent);
            } else if root == current_root && Some(port) > incoming_offset {
                return Some(NeighbourType::Parent);
            } else if seen_roots.insert(root) {
                return Some(NeighbourType::KnownRoot(root));
            } else {
                break;
            }
        }
        path.push(node);
    }
    path.into_iter()
        .find(|n| nodes_with_free_ports.contains(n))
        .map(|n| NeighbourType::NewRoot(n))
}

/// For each node the ports that have not been traversed.
fn free_ports(
    graph: &PortGraph,
    bindings: &HashMap<PGIndexKey, NodeIndex>,
) -> HashMap<NodeIndex, Vec<PortIndex>> {
    // Find the paths that have already been traversed
    // Map from (root, root_offset) to the length of the path traversed
    let mut paths = HashMap::<_, usize>::default();
    let mut roots = HashSet::default();
    for (key, node) in bindings.iter() {
        if let PGIndexKey::AlongPath {
            path_root,
            path_start_port,
            path_length,
        } = key
        {
            let path_id = (*path_root, *path_start_port);
            let prev_len = paths.entry(path_id).or_default();
            *prev_len = (*path_length).max(*prev_len);
        } else {
            roots.insert(node);
        }
    }

    // Traverse each of the traversed paths and remove the ports that have been used
    let mut free_ports = HashMap::default();
    for (&(root, root_offset), &length) in paths.iter() {
        let root = PGIndexKey::PathRoot { index: root };
        let &root_node = bindings
            .get(&root)
            .expect("A PGIndexKey::AlongPath binding references an unbound root");
        for (p1, node, p2) in walk_path(graph, root_node, root_offset).take(length + 1) {
            if roots.contains(&node) {
                continue;
            }
            let free_ports = free_ports
                .entry(node)
                .or_insert_with(|| graph.all_ports(node).collect_vec());
            if let Some(p1) = p1.and_then(|p1| free_ports.iter().position(|&p| p == p1)) {
                free_ports.remove(p1);
            }
            if let Some(p2) = p2.and_then(|p2| free_ports.iter().position(|&p| p == p2)) {
                free_ports.remove(p2);
            }
        }
    }
    free_ports
}

/// All bound nodes that have ports that have not been traversed yet.
///
/// TODO: This is outrageously inefficient.
fn nodes_with_free_ports(
    graph: &PortGraph,
    bindings: &HashMap<PGIndexKey, NodeIndex>,
) -> HashSet<NodeIndex> {
    let free_ports = free_ports(graph, bindings);
    free_ports
        .into_iter()
        .filter(|(_, ports)| !ports.is_empty())
        .map(|(node, _)| node)
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    use insta::assert_debug_snapshot;
    use portgraph::{LinkMut, PortGraph, PortMut};
    use rstest::{fixture, rstest};

    #[fixture]
    fn small_graph() -> PortGraph {
        let mut graph = PortGraph::new();
        let n0 = graph.add_node(2, 2);
        let n1 = graph.add_node(1, 1);
        let n2 = graph.add_node(1, 2);
        let n3 = graph.add_node(1, 2);

        graph.link_nodes(n0, 0, n1, 0).unwrap();
        graph.link_nodes(n1, 0, n2, 0).unwrap();
        graph.link_nodes(n3, 1, n0, 0).unwrap();

        graph
    }

    fn nodes() -> (NodeIndex, NodeIndex, NodeIndex, NodeIndex) {
        (0..4)
            .map(|i| i.try_into().unwrap())
            .collect_tuple()
            .unwrap()
    }

    #[fixture]
    fn bindings() -> HashMap<PGIndexKey, NodeIndex> {
        let (n0, n1, n2, n3) = nodes();
        HashMap::from_iter([
            (PGIndexKey::PathRoot { index: 0 }, n0),
            (
                PGIndexKey::AlongPath {
                    path_root: 0,
                    path_start_port: PortOffset::Outgoing(0),
                    path_length: 1,
                },
                n1,
            ),
            (
                PGIndexKey::AlongPath {
                    path_root: 0,
                    path_start_port: PortOffset::Outgoing(0),
                    path_length: 2,
                },
                n2,
            ),
            (
                PGIndexKey::AlongPath {
                    path_root: 0,
                    path_start_port: PortOffset::Incoming(0),
                    path_length: 1,
                },
                n3,
            ),
        ])
    }

    #[rstest]
    fn test_free_ports(small_graph: PortGraph, bindings: HashMap<PGIndexKey, NodeIndex>) {
        let free_ports = free_ports(&small_graph, &bindings);
        let (_, n1, n2, n3) = nodes();
        let exp_free_ports: HashMap<_, Vec<_>> = [
            (n2, vec![PortOffset::Outgoing(1)]),
            (n3, vec![PortOffset::Incoming(0), PortOffset::Outgoing(0)]),
            (n1, vec![]),
        ]
        .into_iter()
        .map(|(n, offsets)| {
            (
                n,
                offsets
                    .into_iter()
                    .map(|offset| small_graph.port_index(n, offset).unwrap())
                    .collect_vec(),
            )
        })
        .collect();
        assert_eq!(free_ports, exp_free_ports);
    }

    #[rstest]
    fn test_find_root_candidates(small_graph: PortGraph, bindings: HashMap<PGIndexKey, NodeIndex>) {
        let (_, _, n2, n3) = (0..4)
            .map(|i| i.try_into().unwrap())
            .collect_tuple()
            .unwrap();
        let candidates = find_root_candidates(&small_graph, &bindings);
        assert_eq!(HashSet::from_iter(candidates), HashSet::from_iter([n2, n3]));
    }

    #[rstest]
    fn test_spanning_tree(small_graph: PortGraph, mut bindings: HashMap<PGIndexKey, NodeIndex>) {
        let tree = RootSpanningTree::new(&small_graph, &bindings);
        assert_debug_snapshot!("RootSpanningTree_one_root", tree);
        bindings.insert(PGIndexKey::PathRoot { index: 1 }, nodes().2);
        let tree = RootSpanningTree::new(&small_graph, &bindings);
        assert_debug_snapshot!("RootSpanningTree_two_roots", tree);
    }
}
