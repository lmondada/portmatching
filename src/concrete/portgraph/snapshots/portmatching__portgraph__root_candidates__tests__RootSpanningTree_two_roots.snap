---
source: src/portgraph/root_candidates.rs
expression: tree
---
RootSpanningTree {
    nodes: [
        RootSpanningNode {
            neighbours: {
                Incoming(0): NewRoot(
                    NodeIndex(3),
                ),
                Outgoing(0): KnownRoot(
                    1,
                ),
            },
        },
        RootSpanningNode {
            neighbours: {
                Incoming(0): Parent,
            },
        },
    ],
}
